core/src/main/java/ai/timefold/solver/core/impl/score/director/EntityValidator.java

```diff
@@ -0,0 +1,183 @@
+                                            if (!sourceDescriptor.isListVariable()
+                                                    && sourceDescriptor.getEntityDescriptor() == entityDescriptor) {
+                                                out.add(sourceDescriptor);
+                                            }
+                                        }
+                                    }
+                                    return out.toArray(VariableDescriptor[]::new);
+                                });
+                        for (var sourceDescriptor : inverseSourceDescriptors) {
+                            var value = sourceDescriptor.getValue(entity);
+                            valueToInverse
+                                    .computeIfAbsent(value,
+                                            ignored -> CollectionUtils.newIdentityHashSet(DEFAULT_INVERSE_SET_SIZE))
+                                    .add(entity);
+                        }
+                    });
+        }
+    }
+
+    @Override
+    public void accept(Object entity) {
+        var entityDescriptor = solutionDescriptor.findEntityDescriptorOrFail(entity.getClass());
+
+        if (scoreDirector.expectShadowVariablesInCorrectState()) {
+            var inconsistentDescriptor = entityDescriptor.getShadowVariablesInconsistentDescriptor();
+            if (inconsistentDescriptor != null) {
+                if (scoreDirector.expectShadowVariablesInCorrectState()
+                        && inconsistentDescriptor.getValue(entity) == null) {
+                    throw new IllegalStateException(
+                            """
+                                    Shadow variables update is disabled, but the entity (%s) has a null @%s annotated field (%s).
+                                    Maybe enable shadow variable updates?
+                                    """
+                                    .formatted(entity, ShadowVariablesInconsistent.class.getSimpleName(),
+                                            inconsistentDescriptor.getVariableName()));
+                }
+            }
+
+            @SuppressWarnings("unchecked")
+            var inverseRelationDescriptors =
+                    entityClassToInverseDescriptors.computeIfAbsent(entityDescriptor.getEntityClass(),
+                            ignored -> entityDescriptor.getShadowVariableDescriptors()
+                                    .stream()
+                                    .filter(descriptor -> descriptor instanceof InverseRelationShadowVariableDescriptor)
+                                    .toArray(InverseRelationShadowVariableDescriptor[]::new));
+            for (var inverseRelationDescriptor : inverseRelationDescriptors) {
+                assertInverseRelationConsistent(scoreDirector, inverseRelationDescriptor, entity);
+            }
+        }
+
+        if (listVariableDescriptor == null) { // Only basic variables.
```

Comment:
I guess this answers my earlier question - the validator exists so that it can do this. Two follow-ups then:

- Wouldn't it be cleaner, instead of checking for `scoreDirector.expectShadowVariablesInCorrectState` here, to have two different implementations of the validator, with the correct one chosen at instantiation time? One extending the other, perhaps. Or just two different validators, chained together using `andThen(...)`.
- It's still unclear why the shadow var validation exists. Those two points from the previous comment still stand.